src/nvim/search.c

```diff
@@ -4347,121 +4347,288 @@ int linewhite(linenr_T lnum)
+
+      // keep the message even after redraw, but don't put in history
+      msg_hist_off = true;
+      msg_ext_set_kind("search_count");
+      give_warning(msgbuf, false);
+      msg_hist_off = false;
+    }
+}
+
+// Add the search count information to "stat".
+// "stat" must not be NULL.
+// When "recompute" is TRUE always recompute the numbers.
+// dirc == 0: don't find the next/previous match (only set the result to "stat")
+// dirc == '/': find the next match
+// dirc == '?': find the previous match
+static void update_search_stat(int dirc, pos_T *pos, pos_T *cursor_pos,
+                               searchstat_T *stat, int recompute, int maxcount,
+                               long timeout)
+{
+    int             save_ws = p_ws;
+    int             wraparound = false;
+    pos_T           p = (*pos);
+    static pos_T    lastpos = { 0, 0, 0 };
     static int      cur = 0;
     static int      cnt = 0;
+    static int      exact_match = false;
+    static int      incomplete = 0;
+    static int      last_maxcount = SEARCH_STAT_DEF_MAX_COUNT;
     static int      chgtick = 0;
     static char_u   *lastpat = NULL;
     static buf_T    *lbuf = NULL;
-    proftime_T  start;
-#define OUT_OF_TIME 999
+    proftime_T      start;
 
+    memset(stat, 0, sizeof(searchstat_T));
+
+    if (dirc == 0 && !recompute && !EMPTY_POS(lastpos)) {
+      stat->cur = cur;
+      stat->cnt = cnt;
+      stat->exact_match = exact_match;
+      stat->incomplete = incomplete;
+      stat->last_maxcount = last_maxcount;
+      return;
+    }
+    last_maxcount = maxcount;
     wraparound = ((dirc == '?' && lt(lastpos, p))
                   || (dirc == '/' && lt(p, lastpos)));
 
     // If anything relevant changed the count has to be recomputed.
     // STRNICMP ignores case, but we should not ignore case.
     // Unfortunately, there is no STRNICMP function.
+    // XXX: above comment should be "no MB_STRCMP function" ?
     if (!(chgtick == buf_get_changedtick(curbuf)
           && lastpat != NULL  // supress clang/NULL passed as nonnull parameter
           && STRNICMP(lastpat, spats[last_idx].pat, STRLEN(lastpat)) == 0
           && STRLEN(lastpat) == STRLEN(spats[last_idx].pat)
-          && equalpos(lastpos, curwin->w_cursor)
+          && equalpos(lastpos, *cursor_pos)
           && lbuf == curbuf)
-        || wraparound || cur < 0 || cur > 99 || recompute) {
+        || wraparound || cur < 0 || (maxcount > 0 && cur > maxcount)
+        || recompute) {
       cur = 0;
       cnt = 0;
+      exact_match = false;
+      incomplete = 0;
       clearpos(&lastpos);
       lbuf = curbuf;
     }
 
-    if (equalpos(lastpos, curwin->w_cursor) && !wraparound
-        && (dirc == '/' ? cur < cnt : cur > 0)) {
-      cur += dirc == '/' ? 1 : -1;
+    if (equalpos(lastpos, *cursor_pos) && !wraparound
+        && (dirc == 0 || dirc == '/' ? cur < cnt : cur > 0)) {
+      cur += dirc == 0 ? 0 : dirc == '/' ? 1 : -1;
     } else {
+      int done_search = false;
```

Comment:
bool

Reply:
I've changed this and few more to bool . See diff of last commit